### PR TITLE
Elementwise multiplication for Tensor<rank, dim>

### PR DESCRIPTION
**Pull request by anates**
Overloads the operator % for the elementwise multiplication of Tensor-values of same rank. The operator was chosen, after a similar operation can be done for vectors/matrices in the library armadillo. 
If I should use a different operator instead, that can be done, too.

### CODE REVIEW

File: include/deal.II/base/tensor.h [hunk 1]
```diff
@@ -1723,6 +1723,67 @@ inline DEAL_II_ALWAYS_INLINE
   return tmp;
 }
 
+/**
+ * Elementwise multiplication of two tensors of rank 0.
+ *
+ * @relatesalso Tensor
+ */
+
+template <int dim, typename Number, typename OtherNumber>
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: no need for the empty line

File: include/deal.II/base/tensor.h [hunk 2]
```diff
@@ -1723,6 +1723,67 @@ inline DEAL_II_ALWAYS_INLINE
...
+inline DEAL_II_ALWAYS_INLINE
+  Tensor<0, dim, typename ProductType<Number, OtherNumber>::type>
+  operator%(const Tensor<0, dim, Number> &     src1,
+            const Tensor<0, dim, OtherNumber> &src2)
+{
+  Tensor<0, dim, typename ProductType<Number, OtherNumber>::type> tmp(src1);
+
+  for (unsigned int i = 0; i < dim; ++i)
+    tmp[i] *= src2[i];
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: A tensor of rank 0 is just a scalar. How does this work here?

File: include/deal.II/base/tensor.h [hunk 3]
```diff
@@ -1723,6 +1723,67 @@ inline DEAL_II_ALWAYS_INLINE
...
+  return tmp;
+}
+
+/**
+ * Elementwise multiplication of two tensors of rank 1.
+ *
+ * @relatesalso Tensor
+ */
+
+template <int dim, typename Number, typename OtherNumber>
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: same here

File: include/deal.II/base/tensor.h [hunk 4]
```diff
@@ -1723,6 +1723,67 @@ inline DEAL_II_ALWAYS_INLINE
...
+
+/**
+ * Elementwise multiplication of two tensors of general rank.
+ *
+ * @tparam rank The rank of both tensors.
+ *
+ * @relatesalso Tensor
+ */
+
+template <int rank, int dim, typename Number, typename OtherNumber>
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: and here

File: include/deal.II/base/tensor.h [hunk 5]
```diff
@@ -1723,6 +1723,67 @@ inline DEAL_II_ALWAYS_INLINE
...
+inline DEAL_II_ALWAYS_INLINE
+  Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type>
+  operator%(const Tensor<rank, dim, Number> &     src1,
+            const Tensor<rank, dim, OtherNumber> &src2)
+{
+  Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type> tmp(src1);
+
+  for (unsigned int i = 0; i < dim; ++i)
+    tmp[i] = src1[i] % src2[i];
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: nice way to do this recursively :-)

**Comment by anates** [comment]
>     * I'm not a fan of using `operator%`. It's going to be impossible to find this in the documentation because most people will not have an intuition what the operator does. Let's just go with `schur_product` or `elementwise_product`.
> 
I wanted to keep it as simple as possible, i.e. that one is able to write (Tensor<rank, dim> % Tensor<rank, dim>) * Tensor<rank, dim>, for example when looping over cells and adding up gradients. Using an additional function would add additional steps in the code. Would another operator be better?

>     * Ultimately, we will need one or more tests for this functionality. Do you know how to write tests in deal.II?
Yes. I tried to find some tests for operator* or similar operators, but failed. Are there other tests for those operators?

**Comment by bangerth** [comment]
You could do what you want if you used
```
  schur_product(A,B) * C;
```

For tests, there must be tests for functions such as `scalar_product` somewhere in `tests/tensors/`. For example, these might serve as a basis:
```
tests/tensors> grep scalar_product *cc
symmetric_tensor_27.cc:// test scalar_product between tensors and symmetric tensors
symmetric_tensor_27.cc:  deallog << scalar_product(s, s) << std::endl;
symmetric_tensor_27.cc:  AssertThrow(scalar_product(s, s) == s * s, ExcInternalError());
symmetric_tensor_27.cc:  deallog << scalar_product(s, t) << std::endl;
symmetric_tensor_27.cc:  AssertThrow(scalar_product(s, t) == s * symmetrize(t), ExcInternalError());
symmetric_tensor_27.cc:  AssertThrow(scalar_product(s, t) == symmetrize(t) * s, ExcInternalError());
symmetric_tensor_27.cc:  deallog << scalar_product(t, s) << std::endl;
symmetric_tensor_27.cc:  AssertThrow(scalar_product(t, s) == s * symmetrize(t), ExcInternalError());
symmetric_tensor_27.cc:  AssertThrow(scalar_product(t, s) == symmetrize(t) * s, ExcInternalError());
symmetric_tensor_44.cc:  deallog << scalar_product(t2_1, t2_2) << std::endl;
tensor_07.cc:// check scalar_product(Tensor<2,dim>,Tensor<2,dim>)
tensor_07.cc:  deallog << "Constant dim " << dim << '\t' << scalar_product(t, t)
tensor_07.cc:  deallog << "Equal    dim " << dim << '\t' << scalar_product(t, t)
tensor_07.cc:  deallog << "Unequal  dim " << dim << '\t' << scalar_product(s, t)
```

**Comment by anates** [comment]
Test added, not sure if that is enough?

**Comment by anates** [comment]
I added a small description. If that is ok, I will squash the commits.

File: doc/news/changes/minor/20190610RolandRichter [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Added: The function "Schur-product" was added for the Tensor-class, 
+allowing the elementwise multiplication of tensors of general rank
```
Comment (masterleinad) [inline doc/news/changes/minor/20190610RolandRichter]:
```suggestion
allowing the entrywise multiplication of Tensors objects of general rank
```

File: include/deal.II/base/tensor.h [hunk 7]
```diff
@@ -1749,6 +1749,73 @@ DEAL_II_CONSTEXPR inline DEAL_II_ALWAYS_INLINE
   return tmp;
 }
 
+/**
+ * Entrywise multiplication of two tensors of rank 0 (i.e. the multiplication of
```
Comment (masterleinad) [inline include/deal.II/base/tensor.h]:
```suggestion
 * Entrywise multiplication of two Tensor objects of rank 0 (i.e. the multiplication of
```

File: include/deal.II/base/tensor.h [hunk 8]
```diff
@@ -1749,6 +1749,73 @@ DEAL_II_CONSTEXPR inline DEAL_II_ALWAYS_INLINE
...
+ * two scalar values)
```
Comment (masterleinad) [inline include/deal.II/base/tensor.h]:
```suggestion
 * two scalar values).
```

File: include/deal.II/base/tensor.h [hunk 9]
```diff
@@ -1749,6 +1749,73 @@ DEAL_II_CONSTEXPR inline DEAL_II_ALWAYS_INLINE
...
+{
+  Tensor<0, dim, typename ProductType<Number, OtherNumber>::type> tmp(src1);
+
+  tmp *= src2;
+
+  return tmp;
+}
+
+/**
+ * Entrywise multiplication of two tensors of rank 1
```
Comment (masterleinad) [inline include/deal.II/base/tensor.h]:
```suggestion
 * Entrywise multiplication of two Tensor objects of rank 1.
```

File: include/deal.II/base/tensor.h [hunk 10]
```diff
@@ -1749,6 +1749,73 @@ DEAL_II_CONSTEXPR inline DEAL_II_ALWAYS_INLINE
...
+  Tensor<1, dim, typename ProductType<Number, OtherNumber>::type> tmp(src1);
+
+  for (unsigned int i = 0; i < dim; ++i)
+    tmp[i] *= src2[i];
+
+  return tmp;
+}
+
+/**
+ * Entrywise multiplication of two tensors of general rank.
```
Comment (masterleinad) [inline include/deal.II/base/tensor.h]:
```suggestion
 * Entrywise multiplication of two Tensor objects of general rank.
```

File: tests/tensors/tensor_10.cc [hunk 11]
```diff
@@ -0,0 +1,100 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2006 - 2019 by the deal.II authors
```
Comment (masterleinad) [inline tests/tensors/tensor_10.cc]:
```suggestion
// Copyright (C) 2019 by the deal.II authors
```

File: tests/tensors/tensor_10.cc [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  deallog << "Constant rank " << rank << " and dim " << dim << '\t'
+          << schur_product(t, t2) << " compare " << compare << std::endl;
+}
+
+int
+main()
+{
+  std::string   logname = "output";
+  std::ofstream logfile(logname.c_str());
+  deallog.attach(logfile);
```
Comment (masterleinad) [inline tests/tensors/tensor_10.cc]:
```suggestion
  initlog();
```

File: tests/tensors/tensor_10.cc [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+{
+  deallog << "Constant rank " << rank << " and dim " << dim << '\t'
+          << schur_product(t, t2) << " compare " << compare << std::endl;
+}
+
+int
+main()
+{
+  std::string   logname = "output";
+  std::ofstream logfile(logname.c_str());
```
Comment (masterleinad) [inline tests/tensors/tensor_10.cc]:
```suggestion
```

File: tests/tensors/tensor_10.cc [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+               const Tensor<rank, dim> &compare)
+{
+  deallog << "Constant rank " << rank << " and dim " << dim << '\t'
+          << schur_product(t, t2) << " compare " << compare << std::endl;
+}
+
+int
+main()
+{
+  std::string   logname = "output";
```
Comment (masterleinad) [inline tests/tensors/tensor_10.cc]:
```suggestion
```

**Comment by anates** [comment]
I can not squash the commits, after I have a merge in between (updated deal.II after being away for some time). How should I handle that instead?

**Comment by masterleinad** [comment]
What's the problem? You should be able to rebase and do an interactive rebase afterwards, i.e.
```
git rebase upstream/master
git rebase -i upstream/master
```

**Comment by anates** [comment]
Did not know that command, should work now. Thanks!

**Pull request merged**